File: atcoder/abc/abc113/d.cpp
#include "bits/stdc++.h"
using namespace std;

#define FOR(i,a,b) for(int i=(a);i<(b);++i)
#define REP(i,n)   FOR(i,0,n)
#define LL long long

int h,w,k;
LL mod = 1000000007;
LL dp[101][9];

int comb[8] = {1,2,3,5,8,13,21,34};

int main(){
  cin>>h>>w>>k;

  if(w==1){
    cout<<1<<endl;
    return 0;
  }

  REP(i,102){
    REP(j,w+1){
      dp[i][j]=0;
    }
  }

  dp[1][1] = comb[w-2];
  dp[1][2] = comb[max(0,w-3)];
  
  FOR(i,2,h+1){
    FOR(j,1,w+1){
      dp[i][j] = dp[i-1][j] * comb[max(0,j-2)] * comb[max(0,w-j-1)] % mod;
      if(j > 1){
	dp[i][j] += dp[i-1][j-1] * comb[max(0,j-3)] * comb[max(0,w-j-1)] % mod;
      }
      if(j < w){
	dp[i][j] += dp[i-1][j+1] * comb[max(0,j-2)] * comb[max(0,w-j-2)] % mod;
      }
      dp[i][j] %= mod;
    }
  }

  cout<<dp[h][k]%mod<<endl;
  return 0;
}

File: atcoder/abc/abc113/d.hs
import Control.Applicative
import Text.Printf
import Debug.Trace

fib n
  | n >= 1  = fib (n-1) + fib (n-2)
  | n >= -1 = 1
  | otherwise = 0xdeadbeef

mo = 1000000007 :: Integer

comb = fmap fib [0..8]


solve_ 1 w = take w $ [comb !! (w-2),comb !! max 0 (w-3)] ++ zs
  where zs = 0:zs
solve_ h w =
  let dp = solve_ (h-1) w
      c m n q = comb !! max 0 m * comb !! max 0 n * dp !! max 0 q `mod` mo
      left = ((c (w-2) 0 0) + (c (w-3) 0 1)) `mod` mo
      right = ((c (w-2) 0 (w-1)) + (c (w-3) 0 (w-2))) `mod` mo
      mid = [((c (x-2) (w-x-2) (x-1)) + (c (x-1) (w-x-2) x) + (c (x-1) (w-x-3) (x+1))) `mod` mo|x<-[1..w-2]]
--  in trace ("list:" ++ show (left:mid++[right])) (left:mid++[right])
  in left:mid++[right]

solve h w k = solve_ h w !! (k-1)

main = do
  [h,w,k] <- fmap (read :: String -> Int) . words <$> getLine
  if w == 1 then
    printf "1\n"
  else printf "%d\n" $ solve h w k `mod` mo

